sensor: thermistor: improve temperature calculation

or-rhssk opened the pull request:
Tested with STM32F303C8 and Murata NXFT15XH103FA2B060.

Would especially appreciate comments about DTS changes.
In summary: I don't think I've ever seen such a configuration where the thermistor is sandwiched between two fixed value resistors. Is there really a need to require the user to provide two resistance values? Such configuration seems really exotic and is not supported by the driver anyways.

If only the classical scenario should be supported, then it's enough to only require the one fixed resistance value and the rest can be deduced from `connection-type`.

asemjonovs reviewed (dismissed):
Thanks for the improvements.  I modeled this after linux's [implementation](https://www.kernel.org/doc/html/latest/hwmon/ntc_thermistor.html) so that's why I had two resistance values.

or-rhssk commented:
Dropped DT changes since it's quite clear from the linked PRs that Linux compatibility is an important aspect and also to avoid conflicts with #57368.
Only remaining changes are bugfixes.

MaureenHelm commented on drivers/sensor/zephyr_thermistor/zephyr_ntc_thermistor_rt_table.c:
```diff
@@ -130,6 +132,12 @@ int32_t ntc_get_temp_mc(const struct ntc_type *type, uint32_t ohm)
 	temp = ntc_fixp_linear_interpolate(type->comp[low].ohm, type->comp[low].temp_c * 1000,
 					   type->comp[high].ohm, type->comp[high].temp_c * 1000,
 					   ohm);
+	if (temp < min_temp) {
+		temp = min_temp;
+	} else if (temp > max_temp) {
+		temp = max_temp;
+	}
```
can use `CLAMP()` from `util.h`

pdgendt commented on drivers/sensor/zephyr_thermistor/zephyr_ntc_thermistor_rt_table.c:
```diff
@@ -127,6 +153,8 @@ int32_t ntc_get_temp_mc(const struct ntc_type *type, unsigned int ohm)
 	temp = ntc_fixp_linear_interpolate(type->comp[low].ohm, type->comp[low].temp_c * 1000,
 					   type->comp[high].ohm, type->comp[high].temp_c * 1000,
 					   ohm);
+	CLAMP(temp, min_temp, max_temp);
```
You still need to assign it.
```suggestion
	temp = CLAMP(temp, min_temp, max_temp);
```

MaureenHelm reviewed (changes requested):
Please squash the CLAMP changes into `sensor: thermistor: clip interpolated temperature values`

Also need to rebase and resolve the conflicts.

or-rhssk committed:
drivers: sensor: ntc_thermistor: fix out-of-bounds array access

If binary search would return the last array element, index of the higher
point would get a value equal to array size, resulting in out-of-bounds
access.

Signed-off-by: Rihards Skuja <rihards.s@origin-robotics.com>

or-rhssk committed:
drivers: sensor: ntc_thermistor: fix integer overflow

There are multiple places where unsigned integers are freely assigned to
same width but signed integers. In particular, the linear interpolation
method would encounter integer overflow when large but still realistic
parameter values were used.

Fix the overflow and use the right types in other places.

Signed-off-by: Rihards Skuja <rihards.s@origin-robotics.com>

or-rhssk committed:
drivers: sensor: ntc_thermistor: simplify binary search cmp method a bit

Remove redundant comparisons that are already implied by previous ones.

Signed-off-by: Rihards Skuja <rihards.s@origin-robotics.com>

or-rhssk committed:
drivers: sensor: ntc_thermistor: improve resistance calculation algorithm

There were several issues with the previous one:
- it calculated unrealistic resistance values for Vout edge cases (e.g. 0V
  and 3.3V),
- it would try to divide by zero for certain ADC values,
- it assumed that ADC reference voltage is equal to the voltage divider
  input voltage (which, to be fair, will probably be true in most cases).
  Since we're asking user to provide Vin in devicetree anyways, use it.

Signed-off-by: Rihards Skuja <rihards.s@origin-robotics.com>

or-rhssk committed:
drivers: sensor: ntc_thermistor: clarify compensation array size

'n_comp' did not contain array size of 'comp' but of 'comp_table'.

Signed-off-by: Rihards Skuja <rihards.s@origin-robotics.com>

or-rhssk committed:
drivers: sensor: ntc_thermistor: clamp interpolated temperature values

If measured Vout is small or high enough, calculated temperature can get
out of reasonable range. Use the minimum and maximum temperature values
from the generated TR table as lower and upper bounds.

Signed-off-by: Rihards Skuja <rihards.s@origin-robotics.com>

teburd commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -39,10 +39,16 @@ static int ntc_thermistor_sample_fetch(const struct device *dev, enum sensor_cha
 		val_mv = data->raw;
 		res = adc_raw_to_millivolts_dt(&cfg->adc_channel, &val_mv);
-		data->sample_val = val_mv;
+		if (!res) {
```
res == 0

MaureenHelm commented:
Rebase to pick up a fix for the failed Pull Request Assigner check
